Search/ObjectToDocumentConverter.php

```diff
@@ -170,21 +182,31 @@ private function populateDocument($document, $object, $fieldMapping, $prefix = '
                 continue;
             }
 
+            $type = $mapping['type'];
             $value = $this->fieldEvaluator->getValue($object, $mapping['field']);
 
-            if ($value !== null && false === is_scalar($value)) {
-                throw new \InvalidArgumentException(sprintf(
-                    'Search field "%s" resolved to non-scalar value with type "%s". Only scalar (single) values can be indexed.',
-                    $fieldName, gettype($value)
-                ));
+            if ($type !== Field::TYPE_STRING && $type !== Field::TYPE_ARRAY) {
+                $value = $this->converterManager->convert($value, $type);
+
+                $type = is_array($value) ? Field::TYPE_ARRAY : Field::TYPE_STRING;
             }
 
-            if (!is_array($value)) {
+            if ($value !== null && false === is_scalar($value) && false === is_array($value)) {
+                throw new \InvalidArgumentException(
+                    sprintf(
+                        'Search field "%s" resolved to non-scalar value with type "%s". Only scalar (single) values can be indexed.',
```

Comment:
The error message is wrong. And are there still any fields not indexable? Is it possible at all that an object is passed into here?